### PR TITLE
boards: riscv32: Add arty_bm310 board.

### PR DESCRIPTION
**Pull request by vitaly-gaiduk**
Patch contains board configs, serial driver and soc configs for
CloudBEAR BM-310 IP design for Artix-7 35T Arty FPGA.

Signed-off-by: Vitaly Gaiduk <vitaly.gaiduk@cloudbear.ru>

### CODE REVIEW

**Review by galak — CHANGES_REQUESTED** [review]
Please break the change into several commits (SoC, Serial Driver, Board)

File: boards/riscv32/arty_bm310/board.h [hunk 1]
```diff
@@ -0,0 +1,13 @@
+/*
```
Comment (galak) [inline boards/riscv32/arty_bm310/board.h]: Remove board.h if its not needed.

File: dts/riscv32/riscv32-bm310.dtsi [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+			device_type = "cpu";
+			reg = <0>;
+			riscv,isa = "rv32imc";
+			status = "okay";
+			timebase-frequency = <50000000>;
+		};
+	};
+	soc {
+		#address-cells = <1>;
+		#size-cells = <0>;
```
Comment (galak) [inline dts/riscv32/riscv32-bm310.dtsi]: Why is #size-cells 0?  Should be 1.

**Review by galak — CHANGES_REQUESTED** [review]
DT needs cleaning up

**Comment by galak** [comment]
You need board docs

**Comment by galak** [comment]
@pfalcon can you review the serial driver

**Comment by vitaly-gaiduk** [comment]
@galak I will do all needed changes just have two questions:
1) Why does compiler used by CI not know mabi option?
Shippable log:
riscv32-zephyr-elf-gcc: error: unrecognized command line option ‘-mabi=ilp32’
2) What do you mean "DT"?

**Comment by galak** [comment]
> * Why does compiler used by CI not know mabi option?
>   Shippable log:
>   riscv32-zephyr-elf-gcc: error: unrecognized command line option ‘-mabi=ilp32’

@nashif any thoughts/comments on this w/regards to our riscv toolchain. Adding @nategraff-sifive about how this might be working for other riscv platforms/boards.

> * What do you mean "DT"?

Device Tree (DT), sometimes mentioned as "DTS" (device tree source, .dts, etc.)

File: boards/riscv32/arty_bm310/board.h (hunk 1 above)
Comment (vitaly-gaiduk) [reply boards/riscv32/arty_bm310/board.h]: Uses in building console driver

**Comment by pfalcon** [comment]
@vitaly-gaiduk :

Are you sure that designators like uart_cloudbear.c, UART_CLOUDBEAR, etc. are good enough. It's as if you're sure that any design your company would have will use the same UART IP. In other places, you're more specific, e.g. cloudbear-bm310/Kconfig.defconfig.series. Please consider if it's doing the same here.

**Review by pfalcon — DISMISSED** [review]
UART driver seems to be based on old driver API, would need updates at least re: ISR callback.

**Comment by vitaly-gaiduk** [comment]
@pfalcon at this point in time we use one UART IP for any design.

File: boards/riscv32/arty_bm310/board.h (hunk 1 above)
Comment (galak) [reply boards/riscv32/arty_bm310/board.h]: So lets just have the driver include <soc.h> directly.

**Comment by galak** [comment]
> @galak Should I add splitted commits (SoC, Serial Driver) to this pull request or it would be better to create other pull requests?

I think splits commits under one PR is fine for basic SoC, board, and serial driver.  Anything beyond that we should look at additional PRs.  Lets focus on getting the basic in a state that is acceptable for merging.

File: boards/riscv32/arty_bm310/board.h (hunk 1 above)
Comment (vitaly-gaiduk) [reply boards/riscv32/arty_bm310/board.h]: @galak Sorry I didn't notice new notify so I will fix this in new amending :)

**Comment by nategraff-sifive** [comment]
@galak WRT the ABI flag for the toolchain, there's a [pretty good overview of the various options here](https://gnu-mcu-eclipse.github.io/toolchain/riscv/). Since Zephyr hasn't (yet) talked about 64-bit support, we're really only dealing with an -mabi value of "ilp32", "ilp32f", and "ilp32d" depending on whether floats are supported.

For the hifive1, at least, the FE310-G000 doesn't have a floating point unit. This is notated [in the DTS](https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/riscv32/riscv32-fe310.dtsi) with the string `riscv,isa = "rv32imac";`, which doesn't have 'f' or 'd' in the ISA string.

I'm not sure why the SDK's gcc doesn't accept the -mabi option. So far I haven't dug into that part of the build system because I haven't yet run into a need for floats. I can say that we will need the build system to support the flag as soon as we want to allow for RISC-V targets with float support to use it.

File: boards/riscv32/arty_bm310/doc/arty_bm310.rst [hunk 3]
```diff
@@ -0,0 +1,65 @@
+.. _arty_bm310:
+
+CloudBEAR BM-310
+##############
```
Comment (dbkinder) [inline boards/riscv32/arty_bm310/doc/arty_bm310.rst]: Title underline needs to be at least as long as the title, so add a couple more ``##``'s

File: boards/riscv32/arty_bm310/doc/arty_bm310.rst [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+More information about BM-310 can be found on
+`CloudBEAR's website <https://www.cloudbear.ru/bm_310.html>`_.
+
+Programming and debugging
+*************************
+
+Building
+========
+
+Applications for the ``BM-310`` board configuration can be built as usual
```
Comment (dbkinder) [inline boards/riscv32/arty_bm310/doc/arty_bm310.rst]: How about replacing these two sentences with:

    Applications for the BM-310 board can be built as usual
    (see :ref:`build_an_application`), using the ``arty_bm310`` board configuration:

File: boards/riscv32/arty_bm310/doc/arty_bm310.rst [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+to ``arty_bm310``.
+
+.. code-block:: bash
+
+   export BOARD="arty_bm310"
+
+Flashing
+========
+
+In order to upload the application to the device, you'll need OpenOCD
```
Comment (dbkinder) [inline boards/riscv32/arty_bm310/doc/arty_bm310.rst]:
```suggestion
Upload your application to the BM-310 board using OpenOCD
```

File: boards/riscv32/arty_bm310/doc/arty_bm310.rst [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+and GDB with RISC-V support.
+Download and installation instructions can be found in the
+`RISC-V OpenOCD GitHub repository
+<https://github.com/riscv/riscv-openocd>`_ and
+`RISC-V GDB GitHub repository
+<https://github.com/riscv/riscv-binutils-gdb>`_ respectively.
+
+With the necessary tools installed, you can connect to the board using OpenOCD.
```
Comment (dbkinder) [inline boards/riscv32/arty_bm310/doc/arty_bm310.rst]: Change to:

    With the necessary tools installed, establish an OpenOCD connection
    to the board using:

File: boards/riscv32/arty_bm310/doc/arty_bm310.rst [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+To establish an OpenOCD connection run:
+
+.. code-block:: bash
+
+   # assuming that the location of the openocd is in PATH
+   sudo openocd -f your-openocd.cfg
+
+
+Leave it running, and in a different terminal, use GDB to upload the binary to
```
Comment (dbkinder) [inline boards/riscv32/arty_bm310/doc/arty_bm310.rst]: Change to:

    Leave the OpenOCD session running, and in a different terminal,
    use GDB to connect to the device, load the binary, and run the sample:
and start it running:

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: boards/riscv32/arty_bm310/doc/arty_bm310.rst (hunk 4 above)
Comment (vitaly-gaiduk) [reply boards/riscv32/arty_bm310/doc/arty_bm310.rst]: @dbkinder I agree It's better.

**Comment by vitaly-gaiduk** [comment]
@dbkinder I agree with all corrections but I'm waiting for @galak answer about abi option. If it's not planned to add this option soon I could remove 'ilp32' from CMakeList.txt to preserve error.

**Comment by galak** [comment]
> @dbkinder I agree with all corrections but I'm waiting for @galak answer about abi option. If it's not planned to add this option soon I could remove 'ilp32' from CMakeList.txt to preserve error.

In the short term if we can get away w/o the abi option I say lets do that.  Need to see what's involved in support the ABI as we look to migrate the Zephyr SDK toolchains to be built via crosstool-ng in the future.

**Comment by vitaly-gaiduk** [comment]
@galak there is another error in testing:
"Fatal error: unsupported ISA subset i"

**Review by dbkinder — APPROVED** [review]
+1 for doc changes, but there are lots of shippable errors: https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/26522/5/tests

**Comment by vitaly-gaiduk** [comment]
@galak I could remove march option from CMakeLists until this problem with toolchain will be fixed.

**Comment by vitaly-gaiduk** [comment]
@galak I hope that you didn't forget about this PR :) I would like to know what should I do else in this patch for successfully merge.

**Comment by pfalcon** [comment]
@vitaly-gaiduk: In the meantime, I assume you reviewed UART related suggestions and acted on the them if needed, so let me dismiss my -1.

**Comment by vitaly-gaiduk** [comment]
Thanks @pfalcon. It's my inadvertency that I didn't notify you about changes in UART driver - I had updated API signature as mentioned above.

**Comment by galak** [comment]
Feel free to re-open if you want to update this.

**Pull request closed without merging**